## 4.6.1 Combining & Exporting Data

### This script contains the following points:

#### 1. Import the data sets into Jupyter


#### 2. Check the dimensions of the imported dataframes

#### 3. Determine a suitable way to combine the orders_products_combined dataframe with the products data set

#### 4. Confirm the results of the merge using the merge flag

#### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets into Jupyter

In [2]:
# Tell Python to remember a main folder path
path = r'/Users/ashwanisherawat/Downloads/Instacart Basket Analysis/02 Data'

In [3]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'), index_col = False)

### 2. Check the dimensions of the imported dataframes


In [5]:
# Check the output
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [8]:
df_ords_prods_combined.shape

(32434489, 10)

In [9]:
# Drop unnecessary columns
# df_ords_prods_combined = df_ords_prods_combined.drop(['Unnamed: 0.1', 'Unnamed: 0'], axis=1)

In [10]:
# Check the output
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [11]:
# Check the output
df_ords_prods_combined.shape

(32434489, 10)

In [12]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [13]:
df_prods.shape

(49672, 6)

In [14]:
df_ords_prods_merge = pd.merge(
   df_ords_prods_combined,
    df_prods,
    on='product_id',
    how='left'  # Use 'left' to keep all rows from orders_products_combined
)


In [15]:
# Check output
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices  
0            7.0     9.0  
1           16.0    12.5  
2           19.0     4.4  
3           19.0     4.7  
4           17.0     1.0

In [17]:
df_ords_prods_merge.shape

(32435059, 15)

In [23]:
# Create a subset of first million rows

df = df_ords_prods_merge[:1000000]

In [24]:
df_ords_prods_merge['prices'].max()

99999.0

### 4. Confirm the results of the merge using the merge flag

In [18]:
df_ords_prods_merge['_merge'].value_counts()

both          32435059
left_only            0
right_only           0
Name: _merge, dtype: int64

#### The following rows recorded here is 32435059.

### 5. Export the newly created dataframe as ords_prods_merge in a suitable format


In [19]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path,'Prepared Data', 'ords_prods_merge.pkl'))

In [25]:
# Export data to csv
df_ords_prods_merge.to_csv(os.path.join(path,'Prepared Data', 'ords_prods_merge.csv'))